src/main/java/com/bathanh/exercise/basic/string/Ex302.java

```diff
@@ -0,0 +1,19 @@
+package com.bathanh.exercise.basic.string;
+
+public class Ex302 {
+
+    public boolean checkSymmetricString(final String s) {
+        String inputString = s;
+
+        if (inputString.length() == 0 || inputString.length() == 1) {
+            return true;
+        }
+
+        if (inputString.charAt(0) == inputString.charAt(inputString.length() - 1)) {
+            return checkSymmetricString(inputString.substring(1, inputString.length() - 1));
```

Comment:
Thao tác này sẽ bắt buộc Java phải tạo ra String khác hoàn toàn so với String lúc đầu và không hề tối ưu về mặt bộ nhớ. Giả sử input là 1 string gồm 1 triệu ký tự '0' , như vậy

- Mới vào 1'000'000 ký tự
- Đệ quy lần 1 --> 999'998 ký tự
- Đệ quy lần 2 --> 999'996 ký tự
- Đệ quy lần 3 --> 999'994 ký tự
- ...
- Đệ quy lần 599'999 --> 2 ký tự

Java sẽ tạo ra các String với tổng độ dài là 2 + 4 + 6 + 8 + ... + 1'000'000 = 500'000 * 500'001 bằng khoảng 250 * 10^9 ký tự. Mỗi ký tự Java cần 2 bytes để lưu và riêng chỗ đó là 500Gb. Trừ khi dùng siêu máy tính chứ máy tính bình thường là hoàn toàn bất khả thi.

Chưa kể nó cũng mất nửa triệu lần gọi đệ quy, Java sẽ gặp lỗi Stack Overflow

Trong khi đó, 1 triệu ký tự input đầu vào là hoàn toàn bình thường, lưu xuống file text thì chỉ khoảng 2Mb (tuỳ encoding).

Reply:
resolved